src/action_runner/action_runner.cpp

```diff
@@ -33,6 +37,84 @@ int uninstall_msi_action()
     return 0;
 }
 
+namespace fs = std::filesystem;
+
+std::optional<fs::path> copy_self_to_temp_dir()
+{
+    std::error_code error;
+    auto dst_path = fs::temp_directory_path() / "action_runner.exe";
+    fs::copy_file(get_module_filename(), dst_path, fs::copy_options::overwrite_existing, error);
+    if (error)
+    {
+        return std::nullopt;
+    }
+    return std::move(dst_path);
+}
+
+bool install_new_version_stage_1(const bool must_restart = false)
+{
+    std::optional<fs::path> installer;
+    for (auto path : fs::directory_iterator{ updating::get_pending_updates_path() })
+    {
+        if (path.path().native().find(updating::installer_filename_pattern) != std::wstring::npos)
+        {
+            installer.emplace(std::move(path));
+            break;
+        }
+    }
+    if (!installer)
+    {
+        return false;
+    }
+
+    if (auto copy_in_temp = copy_self_to_temp_dir())
+    {
+        // detect if PT was running
+        const bool launch_powertoys = must_restart || FindWindowW(pt_tray_icon_window_class, nullptr) != nullptr;
+
+        std::wstring arguments{ L"-update_now_stage_2 \"" };
+        arguments += installer->c_str();
+        arguments += L"\" \"";
+        arguments += get_module_folderpath();
+        arguments += L"\" ";
+        arguments += launch_powertoys ? L"restart" : L"dont_start";
+        SHELLEXECUTEINFOW sei{ sizeof(sei) };
+        sei.fMask = { SEE_MASK_FLAG_NO_UI | SEE_MASK_NOASYNC };
+        sei.lpFile = copy_in_temp->c_str();
+        sei.nShow = SW_SHOWNORMAL;
+
+        sei.lpParameters = arguments.c_str();
+        return ShellExecuteExW(&sei) == TRUE;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+bool install_new_version_stage_2(std::wstring_view installer_path, std::wstring_view install_path, const bool launch_powertoys)
+{
+    if (MsiInstallProductW(installer_path.data(), nullptr) != ERROR_SUCCESS)
+    {
+        return false;
+    }
+
+    std::error_code _;
+    fs::remove(installer_path, _);
+    if (launch_powertoys)
+    {
+        std::wstring new_pt_path{ install_path };
+        new_pt_path += L"\\PowerToys.exe";
```

Comment:
in the future, i think we should start referencing exe names via a constant file.

Reply:
yes, also since we use Github API releases to get a list of release assets, we should keep the current installer naming convention of `PowerToysSetup-<VERSION>-ARCH.msi`, since we'll need to discriminate between x64/arm in the future. 

perhaps we could have a wiki reminder note for that?

Reply:
Right now we track most stuff like this via issues.  I would label it with Installer and v1.  maybe we should create a new label for Processor Supports